Ensure MySQL tests cleanup connections and raise an exception if they do not

bdraco opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Fixes a few tests not cleaning up connections.

We now raise an exception when a test fails to fully cleanup.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco committed:
Skip db locking test with postgresql/MySQL

https://github.com/home-assistant/core/pull/87756#issuecomment-1423828389

bdraco committed:
Turn back off test_shutdown_before_startup_finishes on MySQL

Was just enabled in https://github.com/home-assistant/core/pull/87756#issuecomment-1423828389
but not yet ready as it is still failing on the CI

bdraco committed:
switch to the guards instead

bdraco committed:
Merge branch 'test_shutdown_before_startup_finishes_mysql' into ii

bdraco committed:
Disable test_migration_in_progress on MySQL

This test was recently enabled on MySQL in
https://github.com/home-assistant/core/pull/87753

Since it the migration is still in progress in the background
when the test ends, it causes deadlock with InnoDB when its dropped
out from under it

bdraco committed:
Merge branch 'disable_test_migration_in_progress' into ii

bdraco committed:
Disable test_entity_ids on MariaDB

Dropping the database after this test will fail on
MySQL and hang forever because it causes an InnoDB deadlock

```
| 2042 | root   | localhost:52698 | NULL               | Query   |   41 | Waiting for table metadata lock | DROP DATABASE `homeassistant-test` |    0.000 |
```

bdraco committed:
Merge branch 'test_entity_ids_innodb_deadlock' into ii

epenet reviewed (approved):
Makes sense, LGTM 👍 
cc @emontnemery

bdraco committed:
better fix, close stale connections

bdraco commented on tests/components/recorder/test_init.py:
```diff
@@ -103,7 +103,6 @@ async def test_shutdown_before_startup_finishes(
 ):
     """Test shutdown before recorder starts is clean."""
-
     if recorder_db_url == "sqlite://":
```
```suggestion

    if recorder_db_url == "sqlite://":
```

bdraco committed:
Update tests/components/recorder/test_init.py

emontnemery commented on tests/conftest.py:
```diff
@@ -1066,7 +1066,26 @@ def _ha_orm_quote(mixed, ident):
+            )
+            for row in connection.execute(query, parameters={"db": db}).fetchall():
+                _LOGGER.warning(
+                    "Killing MySQL connection to temporary database %s", row.id
```
Should we instead fail the test here?

I think recorder tests don't clean up after themselves after the merge of https://github.com/home-assistant/core/pull/84681

After that was merged, there are stray prints from SQLAlchemy:

```
pytest -vv tests/components/recorder

...

 tests/components/recorder/test_init.py::test_connect_args_priority[mysql://user:password@SERVER_IP/DB_NAME?charset=utf8mb4] ✓                                                                          31% ███▏
 tests/components/recorder/test_init.py::test_connect_args_priority[mysql://user:password@SERVER_IP/DB_NAME?blah=bleh&charset=other] ✓                                                                  31% ███▏      INFO:sqlalchemy.engine.Engine:INSERT INTO event_data (hash, shared_data) VALUES (?, ?)
INFO:sqlalchemy.engine.Engine:[cached since 0.002658s ago] (3478486156, '{"domain":"recorder","service":"disable"}')
INFO:sqlalchemy.engine.Engine:INSERT INTO event_data (hash, shared_data) VALUES (?, ?)
INFO:sqlalchemy.engine.Engine:[cached since 0.002985s ago] (2705000516, '{"component":"recorder"}')

INFO:sqlalchemy.engine.Engine:INSERT INTO state_attributes (hash, shared_attrs) VALUES (?, ?)
 tests/components/recorder/test_migrate.py::test_schema_update_calls ✓
```

bdraco replied:
Without it it seems tests can randomly cause the CI to deadlock forever.   I think we should merge this fix now, and try to figure out the root cause when its not blocking the CI.

bdraco replied:
Failing here is a good idea if we can solve the cleanup without leaving the CI broken for too long.

I've got to get some more sleep (only woke up to take out the dog .. and that was a few hours ago) so I'm back off to  bed

epenet committed:
Merge branch 'dev' into ii

epenet commented:
https://github.com/home-assistant/core/pull/87760 was merged, triggering fresh run now.

bdraco committed:
raise if connections active instead

bdraco committed:
Merge branch 'dev' into ii

bdraco committed:
Merge remote-tracking branch 'origin/ii' into ii

bdraco commented on tests/conftest.py:
```diff
@@ -1066,7 +1066,31 @@ def _ha_orm_quote(mixed, ident):
+                raise RuntimeError(
+                    f"Unable to drop database {db} because it is in use by {rows}"
+                )
+            for row in rows:
+                _LOGGER.warning(
+                    "Killing MySQL connection to temporary database %s", row.id
+                )
+                connection.execute(sa.text("KILL :id"), parameters={"id": row.id})
```
```suggestion
```

bdraco committed:
Update tests/conftest.py

bdraco commented on tests/conftest.py:
```diff
@@ -1066,7 +1066,26 @@ def _ha_orm_quote(mixed, ident):
+        made_url = sa.make_url(db_url)
+        db = made_url.database
+        engine = sa.create_engine(db_url)
+        # Kill any open connections to the database before dropping it
```
```suggestion
        # Check for any open connections to the database before dropping it
```

bdraco committed:
Update tests/conftest.py

bdraco commented on tests/conftest.py:
```diff
@@ -1066,7 +1066,26 @@ def _ha_orm_quote(mixed, ident):
+                raise RuntimeError(
+                    f"Unable to drop database {db} because it is in use by {rows}"
+                )
+        engine.dispose()
```
```suggestion
```

bdraco committed:
Update tests/conftest.py

bdraco committed:
fix another test

bdraco committed:
still dispose

bdraco committed:
fix more tests

bdraco committed:
Merge remote-tracking branch 'upstream/dev' into ii

emontnemery committed:
Really ensure recorder test fixture is setup before hass fixture

bdraco committed:
Merge remote-tracking branch 'upstream/recorder_fix_cleanup' into ii

bdraco committed:
fix run history test

bdraco committed:
Merge branch 'dev' into ii

bdraco commented:
Tests are passing locally now so just waiting for the CI to 👍

emontnemery reviewed (approved):
LGTM 👍

bdraco committed:
add missing patch

bdraco committed:
Merge remote-tracking branch 'origin/ii' into ii

bdraco committed:
comment

bdraco committed:
add mocks

bdraco committed:
Merge branch 'dev' into ii

bdraco committed:
make sure event session is still closed if the commit fails

bdraco committed:
Merge remote-tracking branch 'origin/ii' into ii

bdraco committed:
Ensure recorder still shuts down if the final commit fails

If the final commit in _end_session raised, we would
never close the event session which would leave the
connection open leading to an unclean shutdown

bdraco committed:
do not block event loop

bdraco committed:
Merge branch 'ensure_recorder_shutdown' into ii

bdraco committed:
do the session in the executor as well

bdraco committed:
Merge branch 'ensure_recorder_shutdown' into ii

bdraco committed:
safer

bdraco committed:
Merge branch 'ensure_recorder_shutdown' into ii

bdraco committed:
Merge branch 'dev' into ii

bdraco commented on tests/components/recorder/test_init.py:
```diff
@@ -1624,7 +1631,6 @@ async def test_database_lock_timeout(recorder_mock, hass, recorder_db_url):
         return
 
     hass.bus.async_fire(EVENT_HOMEASSISTANT_STOP)
```
```suggestion
    hass.bus.async_fire(EVENT_HOMEASSISTANT_STOP)

```

bdraco committed:
Update tests/components/recorder/test_init.py

bdraco committed:
Merge branch 'dev' into ii
